Sources/BrowserServicesKit/Suggestions/SuggestionProcessing.swift

```diff
@@ -0,0 +1,201 @@
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+/// Class encapsulates the whole ordering and filtering algorithm
+/// It takes query, history, bookmarks, and apiResult as input parameters
+/// The output is instance of SuggestionResult
+final class SuggestionProcessing {
+
+    private var urlFactory: (String) -> URL?
+
+    init(urlFactory: @escaping (String) -> URL?) {
+        self.urlFactory = urlFactory
+    }
+
+    func result(for query: Query,
+                from history: [HistoryEntry],
+                bookmarks: [Bookmark],
+                apiResult: APIResult?) -> SuggestionResult? {
+        let query = query.lowercased()
+
+        let duckDuckGoSuggestions = (try? self.duckDuckGoSuggestions(from: apiResult)) ?? []
+
+        // Get domain suggestions from the DuckDuckGo Suggestions section (for the Top Hits section)
+        let duckDuckGoDomainSuggestions = duckDuckGoSuggestions.compactMap { suggestion -> Suggestion? in
+            guard case let .phrase(phrase) = suggestion, let url = urlFactory(phrase) else {
+                return nil
+            }
+
+            return Suggestion(url: url)
+        }
+
+        // Get best matches from history and bookmarks
+        let allHistoryAndBookmarkSuggestions = historyAndBookmarkSuggestions(from: history,
+                                                                          bookmarks: bookmarks,
+                                                                          query: query)
+
+        // Combine HaB and domains into navigational suggestions and remove duplicates
+        var navigationalSuggestions = allHistoryAndBookmarkSuggestions + duckDuckGoDomainSuggestions
+
+        let maximumOfNavigationalSuggestions = min(
+            Self.maximumNumberOfSuggestions - Self.minimumNumberInSuggestionGroup,
+            query.count * 2)
+        navigationalSuggestions = removeDuplicates(from: navigationalSuggestions,
+                                                   maximum: maximumOfNavigationalSuggestions)
+
+        // Split the Top Hits and the History and Bookmarks section
+        let topHits = Array(navigationalSuggestions.prefix(2))
+        let historyAndBookmarkSuggestions = Array(navigationalSuggestions.dropFirst(2).filter { suggestion in
+            switch suggestion {
+            case .bookmark, .historyEntry:
+                return true
+            default:
+                return false
+            }
+        })
+
+        return makeResult(topHits: topHits,
+                          duckduckgoSuggestions: duckDuckGoSuggestions,
+                          historyAndBookmarks: historyAndBookmarkSuggestions)
+    }
+
+    // MARK: - DuckDuckGo Suggestions
+
+    private func duckDuckGoSuggestions(from result: APIResult?) throws -> [Suggestion]? {
+        // TODO add query as duckduckgo suggestion
```

Comment:
TODO-s should be resolved! 😂

Reply:
Will be soon as a standalone follow up task - https://app.asana.com/0/0/1200372209504780/f